src/v/utils/utf8.h

```diff
@@ -26,8 +26,27 @@ struct default_thrower {
         throw std::runtime_error("Cannot decode string as UTF8");
     }
 };
+struct default_control_thrower {
+    [[noreturn]] [[gnu::cold]] void conversion_error() {
+        throw std::runtime_error("String contains control character");
+    }
+};
 } // namespace
 
+template<typename Thrower>
+requires ExceptionThrower<Thrower>
+inline void validate_no_control(std::string_view s, Thrower&& thrower) {
+    for (auto const& c : s) {
+        if (unlikely((0x00 <= c && c <= 0x1f) || c == 0x7f)) {
+            thrower.conversion_error();
+        }
+    }
```

Comment:
- inline is implicit for template functions
- std::any_of(s.cbegin(), s.cend(), [](char c) { return ...; }))?